### PR TITLE
bugfix: do not mutate initialValue array on multiselect prompt

### LINKED ISSUE #63: [Bug] Initial value array is modified
**Environment**

- OS: macOS
- Node Version: 16.14.0
- Package:  @clack/prompts
- Package Version: 0.4.3

**Describe the bug**
I stumbled on some strange side effects, apparently the array that is passed in to `initialValue` is modified in a `multiselect` if you check in other values than the one you passed in to `initialValue`

**To Reproduce**
1. Create a multi-select prompt
2. Create an appropriate array with e.g. 3 values to use as options.
3. Create another array with 2 of the same values and use as `initialValue`.
4. Select the third value in the prompt.
5. The initialValue array will be modified.

**Expected behavior**
The array should not be mutated.

### PR DESCRIPTION
**Pull request by banjo**
Fixed #63. The array passed into `initialValue` on the `multi-select` prompt is not cloned, so the original array is mutated if a value not selected in the array is selected.

You could use `structuredClone` for this, but it got support from Node 17+, which might not be ideal. I choose the JSON way because it has good support and is not a shallow clone. If I am not mistaken the array should only include strings which won't make it a problem.

### CODE REVIEW

**Review by ulken — CHANGES_REQUESTED** [review]
> You could use structuredClone for this, but it got support from Node 17+, which might not be ideal. I choose the JSON way because it has good support and is not a shallow clone. If I am not mistaken the array should only include strings which won't make it a problem.

I think we should strive to support the active LTS (18) and 16 reaches EOL later this year, so I think using `structuredClone` would be fine. With the library being so new and all, I would have a hard time understanding why anyone starting a new project would go for anything lower.

Right now, primitive values are supported, but that [might change](https://github.com/natemoo-re/clack/issues/44#issuecomment-1435779929), making the JSON parsing/serialization brittle.

For this particular case though, we would get away with a shallow clone, since it's a `push()` which mutates the array. But might not be desirable to prevent future mistakes.

Further more, looking at the implementation of `multiselect`, I've found some more oddities. E.g. for this particular case, I see no reason as to why we keep track of selected options. Just tracking the selected values would be enough. Especially since the UI code maps out the values of the selected options anyway, doing extra, unnecessary work.

I think it would be better to a complete overhaul of the component than to patch individual scratches. I'll have a go tomorrow and get back to you.

**Comment by ulken** [comment]
@banjo if you're affected by this, could you do the job in user land and shallow clone the `initialValue` array until a proper fix lands?

```ts
{
  initialValue: [...initialValue]
}
```

**Comment by banjo** [comment]
> @banjo if you're affected by this, could you do the job in user land and shallow clone the `initialValue` array until a proper fix lands?
> 
> ```ts
> {
>   initialValue: [...initialValue]
> }
> ```

Yeah, that's my current implementation so it's by no means a hurry for my sake.

**Comment by ulken** [comment]
Splendid 👍 

Thanks for being patient and reporting the bug 🐛

**Comment by banjo** [comment]
> Splendid 👍 
> 
> Thanks for being patient and reporting the bug 🐛 

No problem, impressive maintenance from your side. Great job!👌

File: packages/core/src/prompts/multi-select.ts
Comment (natemoo-re) [reply packages/core/src/prompts/multi-select.ts:47]: Yep this seems reasonable to me!

**Comment by natemoo-re** [comment]
Thanks for the PR and for the reviews! I think a shallow clone is okay in this situation, so I'd rather go with that for simplicity.

On mobile rn but will get this merged and released when I'm back at my computer!

**Comment by ulken** [comment]
I have an upcoming PR with (IMO) several improvements (including fixing the issue above) to the implementation of this component. I've made the changes, just have to self-review and write the PR description. 

I'll open it within a few hours, if you want to hold off on merging this.

Of course fine to merge this separately, if you prefer.

**Comment by ulken** [comment]
#65 ready for review.

**Comment by natemoo-re** [comment]
Amazing thank you both! Closing this in favor of #65.

**Pull request closed without merging**
